File: crates/taflow-core/src/stream/rolling_standard_deviation.rs
//! Batch implementation for `rolling_std`.

use super::statistic::*;
use crate::error::{TaError, TaResult};

/// Standard Deviation (STDDEV) — fused single-pass var+sqrt
///
/// Eliminates intermediate var Vec (8MB at 1M bars) by computing
/// Compute the rolling std result for the supplied aligned series.
///
/// # Parameters
///
/// * `input` - Input series or configuration value.
/// * `timeperiod` - Input series or configuration value.
/// * `nbdev` - Input series or configuration value.
///
/// # Returns
///
/// An aligned result with TA-Lib-compatible validation and warm-up values.
use super::rolling_statistics::*;
use super::*;

/// Stateful population standard deviation multiplied by `nbdev`.
#[derive(Debug, Clone)]
/// Persistent Rust state or aligned output type for `RollingStandardDeviation`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct RollingStandardDeviation {
    moments: RollingMoments,
    nbdev: f64,
    value: Option<f64>,
}

impl RollingStandardDeviation {
    /// Computes or updates `new` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn new(period: usize, nbdev: f64) -> TaResult<Self> {
        Ok(Self {
            moments: RollingMoments::new(period)?,
            nbdev,
            value: None,
        })
    }
}

impl StreamingIndicator for RollingStandardDeviation {
    type Output = f64;

    /// Bulk kernel: slice-recurrence sliding moments, bit-identical to
    /// per-bar [`Self::append`] in outputs and post-run state.
    fn extend_slice_into(&mut self, inputs: &[f64], output: &mut Vec<f64>) {
        if inputs.is_empty() {
            return;
        }
        let nbdev = self.nbdev;
        self.value = self.moments.extend_map_into(inputs, output, |variance| {
            stddev_from_variance(variance, nbdev)
        });
    }

    fn append(&mut self, input: f64) -> Option<f64> {
        let nbdev = self.nbdev;
        self.value = self
            .moments
            .append(input)
            .map(|variance| stddev_from_variance(variance, nbdev));
        self.value
    }

    fn value(&self) -> Option<f64> {
        self.value
    }

    fn reset(&mut self) {
        self.moments.reset();
        self.value = None;
    }
}

File: python/taflow/indicators/rolling_standard_deviation.py
"""Canonical native-backed Rolling Standard Deviation adapter."""

from typing import Any
import numpy as np
from .._native import RollingStandardDeviation as _NativeRollingStandardDeviation
from .._series import as_float64_series


class RollingStandardDeviation:
    """Compute STDDEV from required ``values`` with configurable ``nbdev``."""

    def __init__(self, values: Any, timeperiod: int = 14, nbdev: float = 1.0) -> None:
        self._state = _NativeRollingStandardDeviation(timeperiod, nbdev)
        self.extend(values)

    def append(self, value: float) -> "RollingStandardDeviation":
        self._state.append(float(value))
        return self

    def extend(self, values: Any) -> "RollingStandardDeviation":
        self._state.extend(as_float64_series(values))
        return self

    def compute(self) -> np.ndarray:
        return self._state.compute()

    @property
    def value(self) -> float | None:
        return self._state.value

    def reset(self) -> "RollingStandardDeviation":
        self._state.reset()
        return self

    def __len__(self) -> int:
        return len(self._state)


__all__ = ["RollingStandardDeviation"]
